### PR TITLE
Cknieling/add attribute UI

### PR DESCRIPTION
**Pull request by chr11115**
Add attribute UI using spare parameters in the assign node

### CODE REVIEW

File: src/palladio/RuleAttributes.cpp [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+
+		if (attr->getNumParameters() != 0)
+			continue;
+
+		RuleAttribute p;
+		p.fqName = attr->getName();
+		p.niceName = getNiceName(p.fqName);
+		p.mType = attr->getReturnType();
+
+		// TODO: is this correct? import name != rule file name
```
Comment (mistafunk) [inline src/palladio/RuleAttributes.cpp:102]: please resolve the TODO

**Review by mistafunk — CHANGES_REQUESTED** [review]
behavioral issues found in review:
* avoid recooking when opening/closing the parameter folders
* do not set default attribute values (see inline comment)

File: src/palladio/SOPAssign.cpp [hunk 2]
```diff
@@ -146,6 +245,161 @@ bool evaluateDefaultRuleAttributes(const GU_Detail* detail, ShapeData& shapeData
...
+
+					GA_RWHandleI floatHandle(detail->addIntTuple(attrOwner, attributeName, 1));
+					floatHandle.set(0, intValue);
+					break;
+				}
+				case PRM_Type::PRM_BasicType::PRM_BASIC_FLOAT: {
+					const double floatValue = evalFloat(&parm, 0, time);
+
+					GA_RWHandleD floatHandle(detail->addFloatTuple(attrOwner, attributeName, 1));
+					floatHandle.set(0, floatValue);
```
Comment (mistafunk) [inline src/palladio/SOPAssign.cpp:320]: see above re default values

File: src/palladio/SOPAssign.cpp [hunk 3]
```diff
@@ -146,6 +245,161 @@ bool evaluateDefaultRuleAttributes(const GU_Detail* detail, ShapeData& shapeData
...
+					break;
+				}
+				case PRM_Type::PRM_BasicType::PRM_BASIC_STRING: {
+					UT_String stringValue;
+					evalString(stringValue, &parm, 0, time);
+
+					GA_RWHandleS stringHandle(detail->addStringTuple(attrOwner, attributeName, 1));
+					stringHandle.set(0, stringValue);
```
Comment (mistafunk) [inline src/palladio/SOPAssign.cpp:328]: see above re default values

**Review by mistafunk — CHANGES_REQUESTED** [review]
(review continued, prematurely submitted previous review)

there seems to be a mismatch between writing and reading primitive attributes for boolean CGA attributes (int8 storage type). Found with Paris example as well (High_LoD).

File: src/palladio/NodeSpareParameter.h [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "OP/OP_Node.h"
+
+#include "PRM/PRM_Range.h"
+
+using folderVec = std::vector<std::wstring>;
```
[the remark this answers is not in the document]
Comment (chr11115) [reply src/palladio/NodeSpareParameter.h:23]: fixed

File: src/palladio/NodeSpareParameter.cpp [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+
+#include PLD_BOOST_INCLUDE(/container_hash/hash.hpp)
+
+namespace {
+const std::wstring parameterPrefix = L"_palladioAttribute_";
+
+static PRM_Range freeRange = PRM_Range(PRM_RANGE_FREE, 0, PRM_RANGE_FREE);
+
+std::wstring getUniqueIdFromFolderVec(const folderVec& parentFolders) {
+	std::hash<std::wstring> stringHasher;
```
[the remark this answers is not in the document]
Comment (chr11115) [reply src/palladio/NodeSpareParameter.cpp:33]: removed.

File: src/palladio/NodeSpareParameter.cpp [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+	PRM_Name stringParmName(token, label);
+	PRM_Template templateArr[] = {
+		PRM_Template(parmType, 
+		             1, 
+					 &stringParmName, 
+					 &defaultVal,
+					 nullptr,
+					 range),
+		PRM_Template()
```
[the remark this answers is not in the document]
Comment (chr11115) [reply src/palladio/NodeSpareParameter.cpp:96]: fixed

File: src/palladio/AttributeConversion.cpp [hunk 7]
```diff
@@ -524,6 +525,13 @@ std::wstring removeStyle(const std::wstring& n) {
 	return n;
 }
 
+std::wstring removeGroups(const std::wstring& n) {
```
[the remark this answers is not in the document]
Comment (chr11115) [reply src/palladio/AttributeConversion.cpp:528]: fixed.

File: src/palladio/RuleAttributes.cpp (hunk 1 above)
Comment (chr11115) [reply src/palladio/RuleAttributes.cpp:102]: done, should be fine.

File: src/palladio/SOPAssign.cpp [hunk 8]
```diff
@@ -146,6 +245,161 @@ bool evaluateDefaultRuleAttributes(const GU_Detail* detail, ShapeData& shapeData
...
+			switch (currParmType.getBasicType()) {
+				case PRM_Type::PRM_BasicType::PRM_BASIC_ORDINAL: {
+					// only support booleans, i.e. don't store folders
+					if (currParmType.getOrdinalType() != PRM_Type::PRM_OrdinalType::PRM_ORD_TOGGLE)
+						continue;
+
+					const int intValue = evalInt(&parm, 0, time);
+
+					GA_RWHandleI floatHandle(detail->addIntTuple(attrOwner, attributeName, 1));
+					floatHandle.set(0, intValue);
```
[the remark this answers is not in the document]
Comment (chr11115) [reply src/palladio/SOPAssign.cpp:313]: skipping default value parameters and changed handle name to ordinalHandle

File: src/palladio/SOPAssign.cpp (hunk 2 above)
Comment (chr11115) [reply src/palladio/SOPAssign.cpp:320]: fixed

File: src/palladio/NodeSpareParameter.cpp [hunk 9]
```diff
@@ -15,49 +15,55 @@
...
-#include "SOPAssign.h"
+#include "PRM/PRM_SpareData.h"
+#include "UT/UT_VarEncode.h"
 
 #include "BoostRedirect.h"
 #include PLD_BOOST_INCLUDE(/functional/hash.hpp)
 
 namespace {
-const std::wstring parameterPrefix = L"_palladioAttribute_";
+constexpr wchar_t GROUP_SEPARATOR = L'.';
```
Comment (mistafunk) [inline src/palladio/NodeSpareParameter.cpp:29]: this is duplicated from AttributeConversion.h (NameConversion NS)

**Review by mistafunk — CHANGES_REQUESTED** [review]
Good job! Just two tiny code issues left. The behavior issues are resolved.

Feel free to merge afterwards without another review.

File: src/palladio/NodeSpareParameter.cpp [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+#include "OP/OP_Director.h"
+#include "PI/PI_EditScriptedParms.h"
+#include "SOPAssign.h"
+
+#include PLD_BOOST_INCLUDE(/container_hash/hash.hpp)
+
+namespace {
+const std::wstring parameterPrefix = L"_palladioAttribute_";
+
+static PRM_Range freeRange = PRM_Range(PRM_RANGE_FREE, 0, PRM_RANGE_FREE);
```
[the remark this answers is not in the document]
Comment (chr11115) [reply src/palladio/NodeSpareParameter.cpp]: removed.

File: src/palladio/NodeSpareParameter.cpp (hunk 9 above)
Comment (chr11115) [reply src/palladio/NodeSpareParameter.cpp:29]: fixed

File: src/palladio/SOPAssign.cpp [hunk 11]
```diff
@@ -146,6 +245,162 @@ bool evaluateDefaultRuleAttributes(const GU_Detail* detail, ShapeData& shapeData
...
+
+	for (int parmIndex = 0; parmIndex < numParms; ++parmIndex) {
+		const PRM_Parm& parm = getParm(parmIndex);
+
+		if (parm.isSpareParm() && !parm.isDefault()) {
+			const UT_StringHolder attributeName(parm.getToken());
+			const PRM_Type currParmType = parm.getType();
+			GA_AttributeOwner attrOwner = getGroupAttribOwner(GA_GroupType::GA_GROUP_PRIMITIVE);
+
+			switch (currParmType.getBasicType()) {
```
[the remark this answers is not in the document]
Comment (chr11115) [reply src/palladio/SOPAssign.cpp:304]: added default case

**Pull request merged**